[te] Optimize allocation of kernel outputs

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50318 [te] Optimize allocation of kernel outputs**
* #50305 [te] Benchmark comparing fused overhead to unfused

We can skip the dispatcher and go to the device-specific
`at::native::empty_strided` implementation.

Also, unpacking the TensorOptions struct at kernel launch time actually takes a
bit of work, since the optionals are encoded in a bitfield.  Do this upfront
and use the optionals directly at runtime.

Differential Revision: [D25859132](https://our.internmc.facebook.com/intern/diff/D25859132/)

bertmaher committed:
[te] Optimize allocation of kernel outputs

We can skip the dispatcher and go to the device-specific
`at::native::empty_strided` implementation.

Also, unpacking the TensorOptions struct at kernel launch time actually takes a
bit of work, since the optionals are encoded in a bitfield.  Do this upfront
and use the optionals directly at runtime.

Differential Revision: [D25859132](https://our.internmc.facebook.com/intern/diff/D25859132/)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/tensorexpr/codegen.h:
```diff
@@ -66,6 +66,17 @@ class TORCH_API CodeGen {
+      c10::optional<c10::Layout> layout_opt,
+      c10::optional<c10::Device> device_opt,
+      c10::optional<bool> pin_memory_opt) {
+    return at::empty_strided(
```
do we ever use the `at::empty_strided` overload ? is it necessary ?

bertmaher replied:
the SimpleIREval interpreter will use it.  I guess since the interpreter's running on cpu I could just make it call at::native::empty_strided_cpu

eellison reviewed (approved):
Awesome !! I'll have to give your instruction level tracing a try and optimize some other part of the stack! With this and @Krovatkin impending TypeCheck optimization we should be 5-10% faster on a tiny torch.add with the fuser compared to eager 😛

bertmaher committed:
rebase on "[te] Optimize allocation of kernel outputs"


We can skip the dispatcher and go to the device-specific
`at::native::empty_strided` implementation.

Also, unpacking the TensorOptions struct at kernel launch time actually takes a
bit of work, since the optionals are encoded in a bitfield.  Do this upfront
and use the optionals directly at runtime.

Differential Revision: [D25859132](https://our.internmc.facebook.com/intern/diff/D25859132/)

[ghstack-poisoned]

mruberry commented:
Unlanding this stack. It appears to have broken ROCm (linux-bionic-rocm3.10-py3.6-test2). Sample failure:

```
22:09:55   test_kernel_cache_multi_gpu (__main__.TestTEFuser) ... Memory exception on virtual address 0x7fa863400000, node id 5 : Page not present
22:09:56 GPU address 0x7fa863400000, node id 4, size in byte 0x200000
22:09:56 Memory is allocated using hsaKmtAllocMemory
22:09:56 CPU address of the memory is 0x7fa863400000
22:09:56 Memory is mapped to node id: 4 5 6 7 
22:09:56 Memory access fault by GPU node-5 (Agent handle: 0x560b4b94f540) on address 0x7fa863400000. Reason: Page not present or supervisor privilege.
22:09:56 Traceback (most recent call last):
22:09:56   File "test/run_test.py", line 910, in <module>
22:09:56     main()
22:09:56   File "test/run_test.py", line 889, in main
22:09:56     raise RuntimeError(err_message)
22:09:56 RuntimeError: test_jit_fuser_te failed! Received signal: SIGIOT
```
